pkg/uefivarfs/guid.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2014-2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// This code is a somewhat simplified and changed up version of
+// github.com/canonical/go-efilib credits go their authors.
+// It allows interaction with the efivarfs via u-root which means
+// both read and write support.
+
+package uefivarfs
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+)
+
+// GUID is the unique identifier of an efivar
+type GUID [16]byte
+
+// A first segment
+func (guid GUID) A() uint32 {
+	return binary.LittleEndian.Uint32(guid[0:4])
+}
+
+// B second segment
+func (guid GUID) B() uint16 {
+	return binary.LittleEndian.Uint16(guid[4:6])
+}
+
+// C third segment
+func (guid GUID) C() uint16 {
+	return binary.LittleEndian.Uint16(guid[6:8])
+}
+
+// D fourth segment
+func (guid GUID) D() uint16 {
+	return binary.BigEndian.Uint16(guid[8:10])
+}
+
+// E fifth segment
+func (guid GUID) E() [6]uint8 {
+	var out [6]uint8
+	copy(out[:], guid[10:16])
+	return out
+}
+
+// String returns the string representation of a GUID
+func (guid GUID) String() string {
+	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x", guid.A(), guid.B(), guid.C(), guid.D(), guid.E())
+}
+
+// MakeGUID makes a new GUID from the supplied arguments.
+func MakeGUID(a uint32, b, c, d uint16, e [6]uint8) (out GUID) {
+	binary.LittleEndian.PutUint32(out[0:4], a)
+	binary.LittleEndian.PutUint16(out[4:6], b)
+	binary.LittleEndian.PutUint16(out[6:8], c)
+	binary.BigEndian.PutUint16(out[8:10], d)
+	copy(out[10:], e[:])
+	return
+}
+
+// ReadGUID reads a EFI_GUID from the supplied io.Reader.
+func ReadGUID(r io.Reader) (out GUID, err error) {
+	_, err = io.ReadFull(r, out[:])
+	return
+}
+
+var guidRe = regexp.MustCompile(`\{?([[:xdigit:]]{8})-([[:xdigit:]]{4})-([[:xdigit:]]{4})-([[:xdigit:]]{4})-([[:xdigit:]]{12})\}?`)
```

Comment:
All this guid stuff ought to come from our guid package, and we do have a guid.mustcompile function, so I think lines 1-72 should be replaced with uses of our guid package.